ape_hardhat/providers.py

```diff
@@ -0,0 +1,236 @@
+                try:
+                    self._start_process()
+                    break
+                except HardhatSubprocessError as exc:
+                    print("Retrying hardhat subprocess startup:", exc)
+
+        # subprocess should be running and receiving network requests at this point
+        if not (self.process and self.process.poll() is None and self.port):
+            raise HardhatSubprocessError(
+                "Could not start hardhat subprocess on a random port. "
+                "See logs or run `npx hardhat node` or adjust retry strategy configs to "
+                "troubleshoot."
+            )
+
+    @property
+    def uri(self) -> str:
+        uri = getattr(self.config, self.network.ecosystem.name)[self.network.name]["uri"]
+        if uri != DEFAULT_SETTINGS["uri"]:
+            # the user configured their own URI in the project configs, let's use that
+            return uri
+        elif self.port:
+            # the user did not override the default URI, and we have a port
+            # number, so let's build the URI using that port number
+            return f"http://localhost:{self.port}"
+        else:
+            raise RuntimeError("Can't build URI before `connect` is called.")
+
+    @property  # type: ignore
+    def _web3(self):
+        """
+        This property overrides the ``EthereumProvider._web3`` class variable to return our
+        instance variable.
+        """
+        return self._hardhat_web3
+
+    @_web3.setter
+    def _web3(self, value):
+        """
+        Redirect the base class's assignments of self._web3 class variable to our instance variable.
+        """
+        self._hardhat_web3 = value
+
+    def disconnect(self):
+        super().disconnect()
+        if self.process:
+            self.process.kill()
+        self.process = None
+        self.port = None
+
+    def _make_request(self, rpc: str, args: list) -> Any:
+        return self._web3.manager.request_blocking(rpc, args)  # type: ignore
+
+    def set_block_gas_limit(self, gas_limit: int) -> bool:
+        return self._make_request("evm_setBlockGasLimit", [hex(gas_limit)])
+
+    def sleep(self, seconds: int) -> int:
+        return int(self._make_request("evm_increaseTime", [seconds]))
+
+    def mine(self, timestamp: Optional[int] = None) -> str:
+        return self._make_request("evm_mine", [timestamp] if timestamp else [])
+
+    def snapshot(self) -> int:
+        return self._make_request("evm_snapshot", [])
+
+    def revert(self, snapshot_id: int) -> bool:
+        return self._make_request("evm_revert", [snapshot_id])
+
+    def unlock_account(self, address: str) -> bool:
+        return self._make_request("hardhat_impersonateAccount", [address])
+
+
+@plugins.register(plugins.Config)
+def config_class():
+    return HardhatNetworkConfig
+
+
+@plugins.register(plugins.ProviderPlugin)
+def providers():
+    yield "ethereum", "development", HardhatProvider
```

Comment:
oh whoops, I leave these in `__init__.py`

Reply:
Oops, I moved those to __init__ and forgot to remove them from providers. It's removed now.